src/elementPlugins/shortText/src/ApplicationView.tsx

```diff
@@ -9,43 +9,70 @@ const ApplicationView: React.FC<ApplicationViewProps> = ({
   isEditable,
   isRequired,
   allResponses,
+  evaluator,
 }) => {
-  const [validationMessage, setValidationMessage] = useState('')
+  const [validationMessageDisplay, setValidationMessageDisplay] = useState('')
   const [value, setValue] = useState(initialValue?.text)
+  const [isValid, setIsValid] = useState<boolean>()
+  const [responses, setResponses] = useState({
+    thisResponse: undefined,
+    ...allResponses,
+  })
+  const {
+    validation: validationExpression,
+    validationMessage,
+    placeholder,
+    maskedInput,
+  } = templateElement?.parameters
 
-  return (
-    <Form.Input
-      fluid
-      label={templateElement.title}
-      placeholder={templateElement.parameters.placeholder}
-      onChange={onChange(setValue, setValidationMessage, onUpdate)}
-      value={value}
-      disabled={!isEditable}
-      error={
-        validationMessage
-          ? {
-              content: validationMessage,
-              pointing: 'above',
-            }
-          : null
-      }
-    />
-  )
-}
+  useEffect(() => {
+    // Do validation, setIsValid
+    if (validationExpression && responses.thisResponse !== undefined) {
+      evaluator(validationExpression, { objects: [responses] }).then((result: boolean) => {
+        setIsValid(result)
+      })
+    } else setIsValid(true)
+  }, [responses])
+
+  useEffect(() => {
+    setResponses({ thisResponse: value, ...allResponses })
+  }, [value])
+
+  useEffect(() => {
+    if (isValid) setValidationMessageDisplay('')
+    else setValidationMessageDisplay(validationMessage)
+  }, [isValid])
 
-function onChange(updateValue: any, updateValidationMessage: any, onUpdate: any) {
-  return (_: any, { value }: any) => {
-    updateValue(value)
-    // Validation here is just an example.
-    // ideally we use validation condition from templateElement and query evaluator
-    if (value.match(/^[a-zA-Z ]*$/) == null) {
-      updateValidationMessage('Should only have letters and space')
-      onUpdate({ isValid: false })
-    } else {
-      updateValidationMessage('')
-      onUpdate({ value: value, isValid: true })
-    }
+  function handleChange(e: any) {
+    setValue(e.target.value)
   }
+
+  function handleLoseFocus(e: any) {
+    onUpdate({ value: { text: e.target.value }, isValid })
+  }
+
+  return (
+    <>
+      <Form.Input
+        fluid
+        label={templateElement.title}
+        placeholder={placeholder}
+        onChange={handleChange}
+        onBlur={handleLoseFocus}
+        value={value}
+        disabled={!isEditable}
+        type={maskedInput ? 'password' : undefined}
+        error={
+          !isValid && allResponses[templateElement.code]
```

Comment:
&& value ? vs allResponses[templateElement.code]

Reply:
This is my sneaky way of making the validation error notification not show up until you actually *leave* the input box, but only on first visit. After that allResponses[templateElement.code] will have a value in it, so it will show error pop-up on every keystroke. I like it, you may have a different view. Did you actually try it in the UI?

Reply:
I haven't tried it yet, will do before final approval.

I am still not sure what is the difference in validating on keystroke during second entry vs first entry